public/javascripts/ScrabbleBag.js

```diff
@@ -0,0 +1,75 @@
+const _letterBag = new WeakMap()
+const _keys = new WeakMap()
+const _values = new WeakMap()
+const _distinctValues = new WeakMap()
+const _amountsWithLetters = new WeakMap()
+const _errors = new WeakMap()
+
+class ScrabbleBag {
```

Comment:
This class is still doing *a lot* of work for how I conceptualize it (which isn't to say that your approach is wrong, just that we conceive of the problem domain differently.  Here's an approach I would take (I've put comments in to explain the code, though I would most likely omit them in production code):
```js
const LETTER_FREQUENCY = { 'A': 9, /* etc. */, '_': 2 }

class ScrabbleBag {
  constructor() {
    // Make a copy of LETTER_FREQUENCY for use in this instance
    this.letters = Object.assign( {}, LETTER_FREQUENCY )
  }

  playLetters( played ) {
    played.toUpperCase().split( '' ).forEach( letter =>
      if( this.letters[ letter ] === undefined ) {
        console.log( `The letter ${letter} is not a valid letter.` )
      } else if( this.letters[ letter ] === 0 ) {
        console.log( `You entered ${letter} more times than it exists in the bag.` )
      } else {
        this.letters[ letter ]--
      }
    )
  }

  remainingLetters() {
    return this.letters.reduce( (memo, letter) => {
      // Do not add letters with frequency 0 
      if( this.letters[ letter ] && this.letters[ letter ] > 0 ) {
        // Ensure we have created an array for this value
        memo[ this.letters[ letter ] ] = memo[ this.letters[ letter ] ] || []
        memo[ this.letters[ letter ] ].push( letter )
      }

      return memo
    }, {} )
  }
}
```